metrics.go

```diff
@@ -0,0 +1,266 @@
+		})
+		m.registry.MustRegister(m.workerRequestTime[name])
+	}
+
+	if _, ok := m.workerRequestCount[name]; !ok {
+		m.workerRequestCount[name] = prometheus.NewCounter(prometheus.CounterOpts{
+			Namespace: "frankenphp",
+			Subsystem: subsystem,
+			Name:      "worker_request_count",
+		})
+		m.registry.MustRegister(m.workerRequestCount[name])
+	}
+}
+
+func (m *PrometheusMetrics) TotalThreads(num int) {
+	m.totalThreads.Add(float64(num))
+}
+
+func (m *PrometheusMetrics) StartRequest() {
+	m.busyThreads.Inc()
+}
+
+func (m *PrometheusMetrics) StopRequest() {
+	m.busyThreads.Dec()
+}
+
+func (m *PrometheusMetrics) StopWorkerRequest(name string, duration time.Duration) {
+	if _, ok := m.workerRequestTime[name]; !ok {
+		return
+	}
+
+	m.workerRequestCount[name].Inc()
+	m.busyWorkers[name].Dec()
+	m.workerRequestTime[name].Add(duration.Seconds())
+}
+
+func (m *PrometheusMetrics) StartWorkerRequest(name string) {
+	if _, ok := m.busyWorkers[name]; !ok {
+		return
+	}
+	m.busyWorkers[name].Inc()
+}
+
+func (m *PrometheusMetrics) RenameWorker(oldName, newName string) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if _, ok := m.totalWorkers[oldName]; ok {
+		m.totalWorkers[newName] = m.totalWorkers[oldName]
+		delete(m.totalWorkers, oldName)
+	}
+
+	if _, ok := m.busyWorkers[oldName]; ok {
+		m.busyWorkers[newName] = m.busyWorkers[oldName]
+		delete(m.busyWorkers, oldName)
+	}
+
+	if _, ok := m.workerRequestTime[oldName]; ok {
+		m.workerRequestTime[newName] = m.workerRequestTime[oldName]
+		delete(m.workerRequestTime, oldName)
+	}
+
+	if _, ok := m.workerRequestCount[oldName]; ok {
+		m.workerRequestCount[newName] = m.workerRequestCount[oldName]
+		delete(m.workerRequestCount, oldName)
+	}
+}
+
+func (m *PrometheusMetrics) Shutdown() {
+	m.registry.Unregister(m.totalThreads)
+	m.registry.Unregister(m.busyThreads)
+
+	for _, g := range m.totalWorkers {
+		m.registry.Unregister(g)
+	}
+
+	for _, g := range m.busyWorkers {
+		m.registry.Unregister(g)
+	}
+
+	for _, c := range m.workerRequestTime {
+		m.registry.Unregister(c)
+	}
+
+	for _, c := range m.workerRequestCount {
+		m.registry.Unregister(c)
+	}
+
+	m.totalThreads = prometheus.NewCounter(prometheus.CounterOpts{
+		Name: "frankenphp_total_threads",
+		Help: "Total number of PHP threads",
+	})
+	m.busyThreads = prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "frankenphp_busy_threads",
+		Help: "Number of busy PHP threads",
+	})
+	m.totalWorkers = map[string]prometheus.Gauge{}
+	m.busyWorkers = map[string]prometheus.Gauge{}
+	m.workerRequestTime = map[string]prometheus.Counter{}
+	m.workerRequestCount = map[string]prometheus.Counter{}
+
+	m.registry.MustRegister(m.totalThreads)
+	m.registry.MustRegister(m.busyThreads)
+}
+
+func getWorkerNameForMetrics(name string) string {
+	name = regexp.MustCompile(`\W+`).ReplaceAllString(name, "_")
+	name = regexp.MustCompile(`^_+|_+$`).ReplaceAllString(name, "")
+
+	return name
+}
+
+func sanitizeWorkerName(name string) string {
```

Comment:
Couldn't we keep the full path (with special chars escaped if necessary)? This will help distinguishing `/foo/index.php` from `/bar/index.php`, which is quite common.

Reply:
this was used in a test, and I removed it. It is not actually used.